### PR TITLE
[FLINK-1528][Gelly] Added Local Clustering Coefficient Example

### PR DESCRIPTION
**Pull request by balidani**
Hi!

I closed https://github.com/apache/flink/pull/400 and created this instead, because the commit history was messed up.

I have the following exception when running `LocalClusteringCoefficientExample`

    Exception in thread "main" org.apache.flink.api.common.functions.InvalidTypesException: 
    Type of TypeVariable 'K' in 'class flink.graphs.library.LocalClusteringCoefficient
    $NeighborhoodEdgesFunction' could not be determined. 
    This is most likely a type erasure problem. The type extraction currently supports types 
    with generic variables only in cases where all variables in the return type can be deduced 
    from the input type(s).

I have no idea how to fix this though. This is @vasia's old reply:

> The problem is here:
> 
>     DataSet<Tuple2<K, HashSet<K>>> neighborhoods = input.reduceOnEdges(new NeighborhoodEdgesFunction<K>(), EdgeDirection.OUT); 
> 
> and we try to get the return type `Tuple2<K, HashSet<K>>` like this:
> 
>     public TypeInformation<T> getProducedType() {
>          return TypeExtractor.createTypeInfo(EdgesFunction.class, function.getClass(), 2, null, null);
>     }
> 
> Anyone have an idea? Is it because of the nested type parameter in `Tuple2<K, HashSet<K>>` ?

Also, I tried propagating an additional parameter `Q extends HashSet<K>`, but it didn't help.
Does anybody know what's wrong here?

Thanks in advance!

### CODE REVIEW

**Comment by vasia** [comment]
Does anyone have an idea about this? Is there a way to pass the `HashSet<K>` type?

In any case, even if not, I think it doesn't matter in this case.
I believe we shouldn't add this as a library method, as it is a quite naive implementation of local clustering coefficient. I'd prefer if we try to keep the library methods as efficient as possible.

However, I would definitely add this as an example, since it very nicely demonstrates how to use neighborhood methods and `joinWithVertices` (which are missing from the other examples).
So, I would suggest we change this to an example that uses a sample dataset, with e.g. Long ids and also allows file input. What do you think @balidani?

**Comment by balidani** [comment]
Hi!

I agree. I will change the example then :)

**Comment by balidani** [comment]
I made the changes @vasia suggested

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/LocalClusteringCoefficientExample.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+				// Iterate neighbor's neighbors
+				for (Long nn : neighbor.f1.getValue()) {
+					if (vertex.getValue().contains(nn)) {
+						e++;
+					}
+				}
+			}
+
+			// We assume an undirected graph, so we need to divide e here
+			e /= 2;
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/example/LocalClusteringCoefficientExample.java]: I don't get the reasoning behind this.. Do you assume an undirected graph as input? or you convert the input to undirected? in any case, I think we should add a comment in the beginning of the example to define the behavior.

**Comment by vasia** [comment]
Hi @balidani! Thanks for the example :))
I'm a bit confused about the directed / undirected case. I tried testing with both a directed and an undirected input and both my tests failed.. See [here](http://gist.github.com/vasia/ad68268118b6f895cf7c#file-lcc-example-tests) are the test cases I tried.
I suppose the directed case doesn't work because you only consider the out-neighbors, when you should count all. And for the undirected case, I think that the division by 2 you're making is giving a wrong result, because you're not counting edges, just out-neighbors neighbors.

**Comment by balidani** [comment]
Hi @vasia! I fixed the algorithm, now it will convert all edges to a pair of edges and call `distinct` on the edge set. This gives the correct results now. Thanks!

**Comment by vasia** [comment]
Hi @balidani! I think it's still a bit confusing how this example works.
As far as I understand, you expect a directed graph as input, but then you convert it to an undirected one and compute the clustering coefficient of the undirected graph, right?
We should either document this behavior clearly in the example description or compute the result for the input we expect. I personally prefer the second :-)
Also, take a look in the other examples and write a short usage description in the beginning of the example, including the input format that you expect when args are provided.
Let me know if you have questions! Thanks!

**Comment by vasia** [comment]
Hey @balidani!
Would you like to finish this up?
It's not really urgent, but it's almost finished and it'd be a pity to abandon :)
Someone else could also take over of course. Just let us know!

**Comment by balidani** [comment]
Yeah, I should definitely finish this! I'll take a look tonight, sorry about that :)

**Comment by StephanEwen** [comment]
Any progress on this pull request, or should it be closed?

**Comment by vasia** [comment]
@balidani,
I think it'd be better if you close this PR. I don't think we'll add another example after #1000 is merged. I can take over and probably reuse some of your code to add a local clustering coefficient library method. Would that be OK? Thanks!

**Comment by balidani** [comment]
@vasia yes, I'm sorry about not finishing it, but I just did not have the time lately. 

Cheers!

**Comment by vasia** [comment]
Thats' fine :) Thanks for the fast response!

**Pull request closed without merging**